src/DependencyInjection/Compiler/ExtensionCompilerPass.php

```diff
@@ -214,6 +209,32 @@ protected function hasTrait(\ReflectionClass $class, $traitName)
         return $this->hasTrait($parentClass, $traitName);
     }
 
+    private function isExtensionClass(string $type, string $class, string $subject): bool
+    {
+        $classReflection = new \ReflectionClass($class);
+        $subjectReflection = new \ReflectionClass($subject);
+
+        if ('instanceof' === $type) {
+            if ($classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName()) {
+                return true;
+            }
+        } elseif ('implements' === $type) {
+            if ($classReflection->implementsInterface($subject)) {
+                return true;
+            }
+        } elseif ('extends' === $type) {
+            if ($classReflection->isSubclassOf($subject)) {
+                return true;
+            }
+        } elseif ('uses' === $type) {
+            if ($this->hasTrait($classReflection, $subject)) {
+                return true;
+            }
+        }
```

Comment:
```suggestion
        if ('instanceof' === $type && ($classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName())) {
            return true;
        }
        if ('implements' === $type && $classReflection->implementsInterface($subject)) {
            return true;
        }
        if ('extends' === $type && $classReflection->isSubclassOf($subject)) {
            return true;
        }
        if ('uses' === $type && $this->hasTrait($classReflection, $subject)) {
            return true;
        }
```

Reply:
The original solution will be executed much faster if you use `instanceof` for an invalid class, because it will not check every other case.

Reply:
What about a switch on `$type` ?

Reply:
>The original solution will be executed much faster if you use instanceof for an invalid class, because it will not check every other case.

I don't know why, since in the proposed solution each `if` is checking first the type. If the type does not match, the condition is early evaluated as false (the `&&` is omitted) and the next `if` is executed.

Reply:
The original solution will fail fast, e.g. if the type will match. The other `elseif` cases would not be evaluated
```
if ('instanceof' === $type) {
```

While this solution will evaluate every other `elseif` clause:
```
if ('instanceof' === $type && ($classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName())) {
```

Reply:
Same for this place: https://github.com/sonata-project/SonataAdminBundle/pull/6028/files#r403769115

Reply:
Still, 
```
if (something) {
  return true;
}

return false;
```
Is not great

```
return something;
```
Could be better.

WDYT about 
```
switch ($type) {
    case 'instanceof':
        return $classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName();
    case 'implements':
        return $classReflection->implementsInterface($subject);
    case 'extends':
        return $classReflection->isSubclassOf($subject);
    case 'uses':
        return $this->hasTrait($classReflection, $subject);
}
```

Reply:
And for https://github.com/sonata-project/SonataAdminBundle/pull/6028/files#r403769115

You can do 
```
if ('admins' === $type) {
    if ($id === $subject) {
        $extensions = array_merge($extensions, $extensionList);
    }
                    
    continue;
}
                
$class = $this->getManagedClass($admin, $container);
if (class_exists($class) && $this->isExtensionClass($type, $class, $subject)) {
    $extensions = array_merge($extensions, $extensionList);
}
```

It fails as fast as the original version